File: APP-Frontend/src/api/clients/avatarManifest.ts
/**
 * avatar-manifest 域客户端：头像可用动作/表情清单上报（spec enhance-emotion-tts-and-action-presets Task 5.2）。
 *
 * 前端头像 manifest 加载完成后，将可用动作名（motions keys）与表情 id 清单
 * 上报服务端会话级缓存（POST /api/avatar-manifest，覆盖式更新），供后续
 * 服务端提示词注入真实动作名（注入本身归 Task 4，本模块只做上报）。
 *
 * 设计口径：
 * - data 用 snake_case（agent_id）对齐后端契约（与 vision.ts filterFrame 同先例）；
 * - 空 actions 不上报（选择口径：避免空清单覆盖服务端有用缓存；重启后由下次
 *   manifest 就绪的上报自然恢复，无需上报空清单做清理）；
 * - 上报失败静默（console.warn），绝不影响头像加载主流程。
 */
import { request } from '../base';
import type { AvatarManifest } from '../../avatar/types';

export interface ReportAvatarManifestParams {
  agentId: string;
  actions: string[];
  expressions?: string[];
}

export interface ReportAvatarManifestResult {
  status: string;
  agent_id: string;
  actions_count: number;
  expressions_count: number;
  updated_at: string;
}

/** 上报该 agent 的可用动作/表情清单（覆盖式更新；调用方自行处理失败静默） */
export function reportAvatarManifest(
  params: ReportAvatarManifestParams,
): Promise<ReportAvatarManifestResult> {
  const data: Record<string, unknown> = {
    agent_id: params.agentId,
    actions: params.actions,
  };
  // expressions 省略时不下发该字段（后端缺省为空清单）
  if (params.expressions !== undefined) {
    data.expressions = params.expressions;
  }
  return request<ReportAvatarManifestResult>({
    url: '/api/avatar-manifest',
    method: 'post',
    data,
  });
}

/**
 * 从就绪的 AvatarManifest 提取可用动作/表情清单并上报服务端。
 *
 * 上报触发口径（spec「动作清单上报与注入」）：manifest 就绪即上报；
 * 重复加载/切换头像时 manifest 变化会重新调用本函数覆盖服务端缓存。
 *
 * 容错约定：
 * - manifest 缺失或 actions（motions keys）为空时跳过上报，不下发空清单
 *   （避免覆盖服务端有用缓存）；
 * - 上报失败仅 console.warn，不向调用方抛错——绝不影响头像加载主流程。
 */
export async function reportManifestActions(
  manifest: AvatarManifest | null | undefined,
  agentId: string | null | undefined,
): Promise<void> {
  if (!manifest || !agentId) return;
  // 可用动作名 = motions keys；可用表情 id = expressions[].id
  const actions = Object.keys(manifest.motions ?? {});
  if (actions.length === 0) return; // 空清单不上报：避免覆盖服务端有用缓存（口径见文件头注释）
  const expressions = (manifest.expressions ?? []).map((e) => e.id);
  try {
    await reportAvatarManifest({ agentId, actions, expressions });
  } catch (err) {
    // 静默降级：上报失败不影响头像加载主流程（服务端未上报时提示词安全回退）
    console.warn('[avatarManifest] 动作清单上报失败（不影响头像加载）:', err);
  }
}

File: CX-O-SERVER/server/api/routers/avatar_manifest.py
"""头像清单上报 REST 端点（spec enhance-emotion-tts-and-action-presets Task 5.1）。

对外暴露的前端运行时上报/查询接口，挂在 /api 前缀下：

  - POST /api/avatar-manifest                上报可用动作/表情清单（覆盖式更新）
  - GET  /api/avatar-manifest/{agent_id}     查询该 agent 的清单缓存

鉴权口径：与 chat.py 等前端运行时路由一致——不加管理端鉴权依赖
（verify_admin_api_key 仅用于管理操作，上报为前端桌宠窗运行时行为）。

错误映射：
  - 字段缺失 / 类型非法 → 422（Pydantic 校验）
  - 其余内部异常       → 500（错误文案收敛，详情留日志）

业务逻辑委托 server.services.avatar_manifest_registry（进程级内存缓存，
不持久化；服务重启后前端重新上报）。提示词注入（查询消费方）归 Task 4。
"""
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from server.core.logging_config import get_contextual_logger

router = APIRouter(prefix="/avatar-manifest", tags=["avatar-manifest"])
logger = get_contextual_logger(__name__)


class AvatarManifestReport(BaseModel):
    """上报请求体：agent_id 为目标 agent，actions 为可用动作名（motions keys）清单。"""

    agent_id: str = Field(min_length=1, max_length=128, description="目标 Agent id")
    actions: List[str] = Field(default_factory=list, description="可用动作名清单（motions keys）")
    expressions: Optional[List[str]] = Field(default=None, description="可用表情 id 清单（可省略）")


@router.post("")
async def report_manifest(payload: AvatarManifestReport):
    """上报该 agent 的可用动作/表情清单（覆盖式更新）。"""
    from server.services.avatar_manifest_registry import get_avatar_manifest_registry

    try:
        snapshot = get_avatar_manifest_registry().upsert(
            payload.agent_id, payload.actions, payload.expressions
        )
    except Exception as e:  # noqa: BLE001 错误文案收敛：详情留日志
        logger.error(f"头像清单上报失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="头像清单上报处理失败")
    return {
        "status": "success",
        "agent_id": payload.agent_id,
        "actions_count": len(snapshot["actions"]),
        "expressions_count": len(snapshot["expressions"]),
        "updated_at": snapshot["updated_at"],
    }


@router.get("/{agent_id}")
async def get_manifest(agent_id: str):
    """查询该 agent 的清单缓存；未上报时返回空清单与 reported=False（不报错）。"""
    from server.services.avatar_manifest_registry import get_avatar_manifest_registry

    try:
        snapshot = get_avatar_manifest_registry().get(agent_id)
    except Exception as e:  # noqa: BLE001 错误文案收敛：详情留日志
        logger.error(f"头像清单查询失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="头像清单查询处理失败")
    return {
        "agent_id": agent_id,
        "actions": snapshot["actions"],
        "expressions": snapshot["expressions"],
        "updated_at": snapshot["updated_at"],
        "reported": snapshot["updated_at"] is not None,
    }
